service/src/main/java/org/cbioportal/service/impl/CoExpressionServiceImpl.java

```diff
@@ -28,71 +33,125 @@
+        Map<String , List<MolecularData>> molecularDataMapB = molecularDataListB.stream()
+            .collect(Collectors.groupingBy(MolecularData::getStableId));
         
         List<CoExpression> coExpressionList = new ArrayList<>();
-
-        if (queryMolecularDataList == null) {
+        
+        if (!molecularDataMapA.keySet().contains(queryGeneticEntityId)) {
             return coExpressionList;
         }
 
-        List<String> queryValues = queryMolecularDataList.stream().map(g -> g.getValue()).collect(Collectors.toList());
-        for (Integer entrezGeneId : molecularDataMap.keySet()) {
+        List<? extends MolecularData> finalMolecularDataListA = (List<? extends MolecularData>)molecularDataMapA.remove(queryGeneticEntityId);
+        if (molecularDataMapB.get(queryGeneticEntityId) != null) {
+            List<? extends MolecularData> finalMolecularDataListB = (List<? extends MolecularData>)molecularDataMapB.remove(queryGeneticEntityId);
+            if (finalMolecularDataListB == null) {
+                return coExpressionList;
+            }
+        }
+
+        List<String> valuesB = finalMolecularDataListA.stream().map(g -> g.getValue()).collect(Collectors.toList());
+        for (String entityId : molecularDataMapB.keySet()) {
             
-            List<String> values = molecularDataMap.get(entrezGeneId).stream().map(g -> g.getValue())
+            List<String> values = molecularDataMapB.get(entityId).stream().map(g -> g.getValue())
                 .collect(Collectors.toList());
-            List<String> queryValuesCopy = new ArrayList<>(queryValues);
+            List<String> valuesBCopy = new ArrayList<>(valuesB);
 
             List<Integer> valuesToRemove = new ArrayList<>();
-            for (int i = 0; i < queryValuesCopy.size(); i++) {
-                if (!NumberUtils.isNumber(queryValuesCopy.get(i)) || !NumberUtils.isNumber(values.get(i))) {
+            for (int i = 0; i < valuesBCopy.size(); i++) {
+                if (!NumberUtils.isNumber(valuesBCopy.get(i)) || !NumberUtils.isNumber(values.get(i))) {
                     valuesToRemove.add(i);
                 }
             }
 
             for (int i = 0; i < valuesToRemove.size(); i++) {
                 int valueToRemove = valuesToRemove.get(i) - i;
-                queryValuesCopy.remove(valueToRemove);
+                valuesBCopy.remove(valueToRemove);
                 values.remove(valueToRemove);
             }
             
             CoExpression coExpression = new CoExpression();
-            coExpression.setEntrezGeneId(entrezGeneId);
-            Gene gene = genes.get(entrezGeneId).get(0);
-            coExpression.setCytoband(gene.getCytoband());
-            coExpression.setHugoGeneSymbol(gene.getHugoGeneSymbol());
+            coExpression.setGeneticEntityId(entityId);
+            //check if the query gene is a gene or a gene set:
+            if (molecularDataMapB.get(entityId).get(0) instanceof GeneMolecularData) {
```

Comment:
We already know if molecularProfileIdB is a geneset profile or gene profile from the line 87, I don't think you need to check again.